### ISSUE TITLE
Strava API OAuth updates (Old code quits working October 15th, 2019)

### ISSUE THREAD

**Issue by cedricdelpoux** [UnbounDev/node-strava-v3#73]
In case you missed it, Strava is changing OAuth for their API 
https://developers.strava.com/docs/oauth-updates

# Important Dates
## October 15, 2018

New OAuth endpoints for acquiring short-lived access tokens and refresh tokens are available. New scopes are available.

## January 15, 2019

Email address is no longer part of the profile:read_all scope and is removed from the athlete model.

## October 15, 2019

Forever tokens are rejected by the server. Old OAuth endpoints for obtaining forever tokens are removed.

**Comment by xuopled** [comment]
@UnbounDev Do you plan to make some change to this lib about that changes?

**Comment by vlavrynovych** [comment]
@xuopled @TimoSta @UnbounDev 
Hi! Do you still believe that we should work on this repository?
We can easily generate a client code using this: https://developers.strava.com/docs/#client-code
I didn't try it, still using this module, but I've add a ticket to my project to migrate away from this.

**Comment by xuopled** [comment]
I'm not sure to understand how to use this. But if you do or have any examples @vlavrynovych, I would be very interested to know more about this

**Comment by brymon68** [comment]
@xuopled were you ever able to figure out how to implement the new oauth scheme? Do you have any examples you might be able to share?

**Comment by xuopled** [comment]
not yet

**Comment by markstos** [comment]
I plan to stick with this module and intend to implement the OAuth updates before the deadline in October, but a PR is welcome earlier.

I haven't been a big fan of code-gen modules in the past and have already invested some effort into getting my project to work with this Node module, which seems to be working fine for my needs.

However, I have limited time to maintain open source projects so I may eventually switch to the official code-gen solution if that seems like less effort than maintaining this project.

**Comment by ianperrin** [comment]
Not sure if it helps, but I added the new authentication methods in a project that uses this package which seem to do the trick.

Haven’t had chance to create a pull request, but feel free to take a look and see if they can be incorporated into oauth.js

https://github.com/ianperrin/MMM-Strava/blob/develop/strava_api.js

**Comment by markstos** [comment]
@ianperrin Thanks for the reference. A PR would be very welcome. The deadline is coming up and I need to update my own code as well.

**Comment by lavolp3** [comment]
Hi guys,

since I'm (very) actively developing an application using this I took the liberty to include @ianperrin 's code from his MMM-Strava module into my [fork](https://github.com/lavolp3/node-strava-v3/tree/new-oauth-flow).
Will do a local first try with this and then send a PR as a start.
Do I understand it right that there is only a "token exchange" and  a "token refresh" that needs to be added?
The probability is high that I have overseen something...

**Comment by xuopled** [comment]
Nice @lavolp3 Good job 💪🏻

**Comment by markstos** [comment]
I think only a refreshToken method needs to be added. (But I need to test my solution as well) I plan to publish my update later today. I'll review your fork as well to see how it differs from my solution.

**Comment by ianperrin** [comment]
@lavolp3 - great work thank you.

exchangeToken is required to support the oAuth process

refreshToken is required to keep a token alive

I had also overridden getRequestAccessURL to allow for the client secret and id to be passed to it via the args or use the values in the Authenticator. This enables the initial access request.

Hope this helps

**Comment by markstos** [comment]
@ianperrin Doesn't (the poorly documented) `getToken()` meet the need for token exchange already?

**Comment by ianperrin** [comment]
@markstos - I believe the issue with getToken is that to exchange tokens, `grant_type` needs to be included in the args with the value `authorization_code`

**Comment by markstos** [comment]
My proposed 2.0 release is ready on Github, but not yet released. You can review the Release Notes here and see the changes on the master branch now. I'll wait for any comments or feedback for about 24 hours and plan to release this on Thursday or Friday. 

https://github.com/UnbounDev/node-strava-v3/releases

**Comment by markstos** [comment]
@ianperrin As I read the API docs now, grant_type must be set to `authorization_code` or `refresh_token`. Currently `getToken()` sets neither. I'll update it to always set `authorization_code` which will resolve the issue without adding a third method. Working on that today.

**Comment by ianperrin** [comment]
@markstos - great, I'll see if I can give it a test, when I can

I think you'll still need to either include exchangeToken or modify getToken to allow grant_type to be passed ( https://developers.strava.com/docs/authentication/#token-exchange )

perhaps the following will work

````
oauth.exchangeToken = (code, done) => {
  const options = {
    url: 'https://www.strava.com/oauth/token',
    method: 'POST',
    json: true,
    simple: true,
    qs: {
      code : code,
      client_id : authenticator.getClientId(),
      client_secret : authenticator.getClientSecret(),
      grant_type : "authorization_code",
    }
  }
  return Promise.resolve(request(options)).asCallback(done)
}
````

**Comment by ianperrin** [comment]
Sorry - comments crossed over the wire.

I think getToken can be modified, but not sure if it's needed elsewhere without the grant_type.

Happy with a 3rd method or a modified getToken - will leave it to you to decide which works best

;)

**Comment by lavolp3** [comment]
I have just seen that a lot has happened in the recent 24 hours :-)
I have just sent a PR that might in the meantime have become unnecessary.
I'll keep it open for you @markstos and @ianperrin to see what changes I have made. It works with my MMM-Strava fork.
However, as said, if the issue is already solved with your changes, I'm happy as well and will close the PR.

**Comment by markstos** [comment]
I've pushed a new commit to update `getToken()`, including some internal docs for it and updating the testing of it. I found a couple other spec compliance things to clean up besides setting the grant_type. More details in the commit message here:

https://github.com/UnbounDev/node-strava-v3/commit/c3049599640109b23bd17c9c349e4421e55626cb

**Comment by markstos** [comment]
This was fixed in the 2.0.0 release.
